src/replication/adapter/awsecr/auth.go

```diff
@@ -0,0 +1,146 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsecr
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	awsecrapi "github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/goharbor/harbor/src/common/http/modifier"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/common/utils/registry"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// Credential ...
+type Credential modifier.Modifier
+
+// Implements interface Credential
+type awsAuthCredential struct {
+	region       string
+	accessKey    string
+	accessSecret string
+	insecure     bool
+
+	cacheToken   *cacheToken
+	cacheExpired *time.Time
+}
+
+type cacheToken struct {
+	endpoint string
+	user     string
+	password string
+	host     string
+}
+
+// DefaultCacheExpiredTime is expired timeout for aws auth token
+const DefaultCacheExpiredTime = time.Hour * 1
+
+func (a *awsAuthCredential) Modify(req *http.Request) error {
+	if !a.isTokenValid() {
+		endpoint, user, pass, err := a.getAuthorization()
+
+		if err != nil {
+			return err
+		}
+		u, err := url.Parse(endpoint)
+		if err != nil {
+			return err
+		}
+		a.cacheToken = &cacheToken{}
+		a.cacheToken.host = u.Host
+		a.cacheToken.user = user
+		a.cacheToken.password = pass
+		a.cacheToken.endpoint = endpoint
+		t := time.Now().Add(DefaultCacheExpiredTime)
+		a.cacheExpired = &t
+	}
+	req.Host = a.cacheToken.host
+	req.URL.Host = a.cacheToken.host
+	req.SetBasicAuth(a.cacheToken.user, a.cacheToken.password)
+	return nil
+}
+
+func (a *awsAuthCredential) getAuthorization() (string, string, string, error) {
+	log.Infof("Aws Ecr getAuthorization %s", a.accessKey)
+	cred := credentials.NewStaticCredentials(
+		a.accessKey,
+		a.accessSecret,
+		"")
+	sess := session.Must(session.NewSession(&aws.Config{
```

Comment:
Seems the `must` will panic the process, right?  The error can be wrapped and returned.